water-effect/particle.js

```diff
@@ -0,0 +1,1021 @@
+      const maxEmit = (Math.floor(this.currentSpeed * 10 + 1) * 5);
+      for (let i = 0; i < maxEmit; i ++) {
+        localVector3.set(this.player.position.x, this.player.position.y, this.player.position.z);
+        if (scalesAttribute.getX(i) <= 0) {
+          if (this.currentSpeed > 0.1) {
+            localVector3.x += (Math.random() - 0.5) * 0.5;
+            localVector3.y += -0.2 - (Math.random()) * 0.5;
+            localVector3.z += (Math.random() - 0.5) * 0.5;
+            this.bubble.info.velocity[i].x = -this.playerDir.x * 0.005;
+            this.bubble.info.velocity[i].y = 0.0025 + Math.random() * 0.0025;
+            this.bubble.info.velocity[i].z = -this.playerDir.z * 0.005;
+          }
+          else { // when stop moving
+            localVector3.x += -this.playerDir.x * 0.25 + (Math.random() - 0.5) * 0.5;
+            localVector3.z += -this.playerDir.z * 0.25 + (Math.random() - 0.5) * 0.5;
+            const avatarHeight = this.player.avatar.height;
+            localVector3.y -= avatarHeight * 0.6 + (Math.random()) * 0.2;
+            this.bubble.info.velocity[i].x = 0;
+            this.bubble.info.velocity[i].y = 0.0025 + Math.random() * 0.0025;
+            this.bubble.info.velocity[i].z = 0;
+          }
+          if (localVector3.y > this.waterSurfaceHeight) // avoid bubble fly above water
+            localVector3.y = this.waterSurfaceHeight;
+          positionsAttribute.setXYZ(i, localVector3.x, localVector3.y, localVector3.z);
+          
+          this.bubble.info.offset[i] = Math.floor(Math.random() * 29);
+          this.bubble.info.maxLife[i] = (50 + Math.random() * 50);
+          this.bubble.info.livingTime[i] = 0;
+          scalesAttribute.setX(i, Math.random());
+        }
+      }
+    }
+    positionsAttribute.needsUpdate = true;
+    scalesAttribute.needsUpdate = true;
+  }
+
+  updateBubble() {
+    if (this.bubble) {
+      const particleCount = this.bubble.info.particleCount;
+      const offsetAttribute = this.bubble.geometry.getAttribute('offset');
+      const positionsAttribute = this.bubble.geometry.getAttribute('positions');
+      const scalesAttribute = this.bubble.geometry.getAttribute('scales');
+      for (let i = 0; i < particleCount; i++) {
+        if (scalesAttribute.getX(i) > 0) {
+          if (positionsAttribute.getY(i) >= this.waterSurfaceHeight) {
+            this.bubble.info.velocity[i].y = 0;
+          }
+          positionsAttribute.setXYZ(  
+            i, 
+            positionsAttribute.getX(i) + this.bubble.info.velocity[i].x,
+            positionsAttribute.getY(i) + this.bubble.info.velocity[i].y,
+            positionsAttribute.getZ(i) + this.bubble.info.velocity[i].z
+          );
+
+          // handle texture offset
+          this.bubble.info.livingTime[i] ++;
+          if (this.bubble.info.livingTime[i] % 2 === 0)
+            this.bubble.info.offset[i] ++;
+          if (this.bubble.info.offset[i] >= 30)
```

Comment:
These values should be variables, constants at the top of the file, so that we can change them easily